core/common/src/main/java/alluxio/util/ShellUtils.java

```diff
@@ -193,6 +194,112 @@ private String run() throws ExitCodeException, IOException {
         process.destroy();
       }
     }
+
+    /**
+     * Runs a command and returns its stdout, stderr and exit code.
+     * No matter it succeeds or not.
+     *
+     * @return {@link CommandReturn} object representation of stdout, stderr and exit code
+     */
+    protected CommandReturn runTolerateFailure() throws IOException {
+      Process process = new ProcessBuilder(mCommand).start();
+      CommandReturn cr = null;
+
+      try (BufferedReader inReader =
+                   new BufferedReader(new InputStreamReader(process.getInputStream()));
+           BufferedReader errReader =
+                   new BufferedReader(new InputStreamReader(process.getErrorStream()))) {
+        // read the output of the command
+        StringBuilder stdout = new StringBuilder();
+        String outLine = inReader.readLine();
+        while (outLine != null) {
+          stdout.append(outLine);
+          stdout.append("\n");
+          outLine = inReader.readLine();
+        }
+
+        // read the stderr of the command
+        StringBuilder stderr = new StringBuilder();
+        String errLine = errReader.readLine();
```

Comment:
will reading err after fully reading stdout work? I'm not sure how these work, but do both streams "end" at the same time?

Reply:
This is a great question! In short it's wrong to read the 2 streams sequentially like what I did here. The correct way is to read them in two separated threads. 

On 2nd thought, the benefit of having stderr separated from stdout is out-weighted by the overhead of having another thread reading stderr (stdout can be read in the same thread cuz we are blocking anyway). So I'm now redirecting stderr into stdout. The corresponding `stderr` field in `CommandReturn` class has been removed cuz there's currently no need of it.